File: apiendpoint/app/core/alibaba_config.py
"""
Alibaba Cloud Configuration for FastAPI Query Service
Handles RDS MySQL, OSS, and other Alibaba Cloud services
"""

import os
from typing import Optional
from functools import lru_cache
from dotenv import load_dotenv

# Load .env.alibaba file
load_dotenv(os.path.join(os.path.dirname(__file__), '../../.env.alibaba'))


class AlibabaCloudConfig:
    """Alibaba Cloud Configuration"""
    
    # ==================================================
    # ALIBABA CLOUD CREDENTIALS
    # ==================================================
    access_key_id: str = os.getenv("ALIBABA_ACCESS_KEY_ID", "")
    access_key_secret: str = os.getenv("ALIBABA_ACCESS_KEY_SECRET", "")
    region: str = os.getenv("ALIBABA_REGION", "ap-southeast-1")
    
    # ==================================================
    # APSARADB RDS MYSQL DATABASE
    # ==================================================
    db_host: str = os.getenv("DB_HOST", "localhost")
    db_port: int = int(os.getenv("DB_PORT", 3306))
    db_user: str = os.getenv("DB_USER", "root")
    db_password: str = os.getenv("DB_PASSWORD", "")
    db_name: str = os.getenv("DB_NAME", "wms_manufacture")
    
    # Connection pool settings
    db_connection_limit: int = int(os.getenv("DB_CONNECTION_LIMIT", 10))
    db_ssl_enabled: bool = os.getenv("DB_SSL_ENABLED", "false").lower() == "true"
    db_ssl_ca_path: Optional[str] = os.getenv("DB_SSL_CA_PATH")
    
    # ==================================================
    # ALIBABA OSS (OBJECT STORAGE SERVICE)
    # ==================================================
    oss_endpoint: str = os.getenv("OSS_ENDPOINT", "")
    oss_bucket: str = os.getenv("OSS_BUCKET_NAME", "")
    oss_region: str = os.getenv("OSS_REGION", "ap-southeast-1")
    
    # ==================================================
    # ALIBABA MESSAGE QUEUE (Optional)
    # ==================================================
    mq_endpoint: str = os.getenv("MQ_ENDPOINT", "")
    mq_queue_name: str = os.getenv("MQ_QUEUE_NAME", "")
    mq_topic_name: str = os.getenv("MQ_TOPIC_NAME", "")
    mq_consumer_group: str = os.getenv("MQ_CONSUMER_GROUP", "")
    mq_username: Optional[str] = os.getenv("MQ_USERNAME")
    mq_password: Optional[str] = os.getenv("MQ_PASSWORD")
    
    # ==================================================
    # SERVICE CONFIGURATION
    # ==================================================
    query_service_port: int = int(os.getenv("QUERY_SERVICE_PORT", 2025))
    query_service_host: str = os.getenv("QUERY_SERVICE_HOST", "0.0.0.0")
    
    # ==================================================
    # VPC & NETWORKING
    # ==================================================
    vpc_id: str = os.getenv("VPC_ID", "")
    security_group_id: str = os.getenv("SECURITY_GROUP_ID", "")
    ecs_instance_id: str = os.getenv("ECS_INSTANCE_ID", "")
    ecs_private_ip: str = os.getenv("ECS_PRIVATE_IP", "")
    
    # ==================================================
    # BACKUP CONFIGURATION
    # ==================================================
    backup_enabled: bool = os.getenv("BACKUP_ENABLED", "false").lower() == "true"
    backup_frequency: str = os.getenv("BACKUP_FREQUENCY", "daily")
    backup_time: str = os.getenv("BACKUP_TIME", "02:00")
    backup_retention_days: int = int(os.getenv("BACKUP_RETENTION_DAYS", 30))
    backup_location: str = os.getenv("BACKUP_LOCATION", "")
    
    # ==================================================
    # SECURITY
    # ==================================================
    jwt_secret: str = os.getenv("JWT_SECRET", "")
    jwt_expiry: str = os.getenv("JWT_EXPIRY", "24h")
    
    # ==================================================
    # ENVIRONMENT
    # ==================================================
    python_env: str = os.getenv("PYTHON_ENV", "development")
    
    # ==================================================
    # SQLALCHEMY DATABASE URL
    # ==================================================
    @property
    def database_url(self) -> str:
        """Generate SQLAlchemy database URL for Alibaba RDS MySQL"""
        if self.db_ssl_enabled and self.db_ssl_ca_path:
            return f"mysql+pymysql://{self.db_user}:{self.db_password}@{self.db_host}:{self.db_port}/{self.db_name}?ssl_ca={self.db_ssl_ca_path}&ssl_verify_cert=true"
        return f"mysql+pymysql://{self.db_user}:{self.db_password}@{self.db_host}:{self.db_port}/{self.db_name}"
    
    # ==================================================
    # CHECK IF CONFIGURED
    # ==================================================
    def is_configured(self) -> bool:
        """Check if Alibaba Cloud is properly configured"""
        return (
            self.access_key_id != "" and 
            self.access_key_id != "your_access_key_id_here" and
            self.access_key_secret != "" and
            self.access_key_secret != "your_access_key_secret_here" and
            self.db_host != "" and
            self.db_host != "rm-xxxxx.mysql.rds.aliyuncs.com"
        )
    
    def get_info(self) -> dict:
        """Get configuration info for display"""
        return {
            "region": self.region,
            "db_host": self.db_host,
            "db_port": self.db_port,
            "db_name": self.db_name,
            "oss_endpoint": self.oss_endpoint,
            "oss_bucket": self.oss_bucket,
            "configured": self.is_configured(),
            "environment": self.python_env,
        }


# Create singleton instance
@lru_cache()
def get_alibaba_config() -> AlibabaCloudConfig:
    """Get Alibaba Cloud configuration (cached)"""
    return AlibabaCloudConfig()


# For backward compatibility
alibaba_config = get_alibaba_config()

File: apiendpoint/node/config/alibaba.config.js
/**
 * Alibaba Cloud Configuration
 * Handles RDS MySQL, OSS, and other Alibaba Cloud services
 */

const path = require('path');
require('dotenv').config({ path: path.join(__dirname, '../../.env.alibaba') });

module.exports = {
  // ==================================================
  // ALIBABA CLOUD CREDENTIALS
  // ==================================================
  credentials: {
    accessKeyId: process.env.ALIBABA_ACCESS_KEY_ID,
    accessKeySecret: process.env.ALIBABA_ACCESS_KEY_SECRET,
    region: process.env.ALIBABA_REGION || 'ap-southeast-1',
  },

  // ==================================================
  // APSARADB RDS MYSQL CONNECTION
  // ==================================================
  mysql: {
    host: process.env.DB_HOST,
    port: parseInt(process.env.DB_PORT) || 3306,
    user: process.env.DB_USER,
    password: process.env.DB_PASSWORD,
    database: process.env.DB_NAME,
    
    // Connection pool
    waitForConnections: process.env.DB_WAIT_FOR_CONNECTIONS === 'true',
    connectionLimit: parseInt(process.env.DB_CONNECTION_LIMIT) || 10,
    queueLimit: parseInt(process.env.DB_QUEUE_LIMIT) || 0,
    
    // SSL/TLS Configuration (if enabled)
    ssl: process.env.DB_SSL_ENABLED === 'true' ? {
      ca: process.env.DB_SSL_CA_PATH,
      rejectUnauthorized: false,
    } : false,
    
    // Sequelize Options for ORM
    sequelize: {
      dialect: 'mysql',
      host: process.env.DB_HOST,
      port: parseInt(process.env.DB_PORT) || 3306,
      username: process.env.DB_USER,
      password: process.env.DB_PASSWORD,
      database: process.env.DB_NAME,
      logging: process.env.NODE_ENV === 'production' ? false : console.log,
      pool: {
        max: parseInt(process.env.DB_CONNECTION_LIMIT) || 10,
        min: 0,
        acquire: 30000,
        idle: 10000,
      },
    },
  },

  // ==================================================
  // ALIBABA OSS (OBJECT STORAGE SERVICE)
  // ==================================================
  oss: {
    region: process.env.OSS_REGION || 'ap-southeast-1',
    accessKeyId: process.env.OSS_ACCESS_KEY_ID || process.env.ALIBABA_ACCESS_KEY_ID,
    accessKeySecret: process.env.OSS_ACCESS_KEY_SECRET || process.env.ALIBABA_ACCESS_KEY_SECRET,
    bucket: process.env.OSS_BUCKET_NAME,
    endpoint: process.env.OSS_ENDPOINT,
  },

  // ==================================================
  // ALIBABA MESSAGE QUEUE (Optional)
  // ==================================================
  mq: {
    enabled: !!process.env.MQ_ENDPOINT,
    endpoint: process.env.MQ_ENDPOINT,
    queueName: process.env.MQ_QUEUE_NAME,
    topicName: process.env.MQ_TOPIC_NAME,
    consumerGroup: process.env.MQ_CONSUMER_GROUP,
    username: process.env.MQ_USERNAME,
    password: process.env.MQ_PASSWORD,
  },

  // ==================================================
  // SERVICE CONFIGURATION
  // ==================================================
  services: {
    commandService: {
      port: parseInt(process.env.COMMAND_SERVICE_PORT) || 3108,
      host: process.env.COMMAND_SERVICE_HOST || '0.0.0.0',
    },
    queryService: {
      port: parseInt(process.env.QUERY_SERVICE_PORT) || 2025,
      host: process.env.QUERY_SERVICE_HOST || '0.0.0.0',
    },
  },

  // ==================================================
  // VPC & NETWORKING
  // ==================================================
  vpc: {
    vpcId: process.env.VPC_ID,
    securityGroupId: process.env.SECURITY_GROUP_ID,
    ecsInstanceId: process.env.ECS_INSTANCE_ID,
    ecsPrivateIp: process.env.ECS_PRIVATE_IP,
  },

  // ==================================================
  // LOG SERVICE (Optional)
  // ==================================================
  logService: {
    enabled: !!process.env.LOG_SERVICE_ENDPOINT,
    endpoint: process.env.LOG_SERVICE_ENDPOINT,
    project: process.env.LOG_SERVICE_PROJECT,
    logstore: process.env.LOG_SERVICE_LOGSTORE,
    accessKeyId: process.env.ALIBABA_ACCESS_KEY_ID,
    accessKeySecret: process.env.ALIBABA_ACCESS_KEY_SECRET,
  },

  // ==================================================
  // BACKUP CONFIGURATION
  // ==================================================
  backup: {
    enabled: process.env.BACKUP_ENABLED === 'true',
    frequency: process.env.BACKUP_FREQUENCY || 'daily',
    time: process.env.BACKUP_TIME || '02:00',
    retentionDays: parseInt(process.env.BACKUP_RETENTION_DAYS) || 30,
    location: process.env.BACKUP_LOCATION,
  },

  // ==================================================
  // SECURITY
  // ==================================================
  security: {
    jwtSecret: process.env.JWT_SECRET,
    jwtExpiry: process.env.JWT_EXPIRY || '24h',
  },

  // ==================================================
  // ENVIRONMENT
  // ==================================================
  environment: process.env.NODE_ENV || 'development',
};
